src/sql/workbench/contrib/notebook/test/electron-browser/notebookUtils.test.ts

```diff
@@ -3,86 +3,184 @@
+		assert.strictEqual(result, null);
+
+		result = tryMatchCellMagic('%%sql');
+		assert.strictEqual(result, 'sql');
+
+		result = tryMatchCellMagic('%%');
+		assert.strictEqual(result, null);
+
+		result = tryMatchCellMagic('%% sql');
+		assert.strictEqual(result, null);
 	});
 
-	test('Should extract server name when no database specified', async function (): Promise<void> {
-		let serverName = getServerFromFormattedAttachToName('serverName');
-		let databaseName = getDatabaseFromFormattedAttachToName('serverName');
-		assert.equal(serverName, 'serverName');
-		assert.equal(databaseName, '');
+	test('asyncForEach Test', async function (): Promise<void> {
+		let totalResult = 0;
+		await asyncForEach([1, 2, 3, 4], async (value) => {
+			totalResult += value;
+		});
+		assert.strictEqual(totalResult, 10);
 	});
 
-	test('Should extract server and database name', async function (): Promise<void> {
-		let serverName = getServerFromFormattedAttachToName('serverName (databaseName)');
-		let databaseName = getDatabaseFromFormattedAttachToName('serverName (databaseName)');
-		assert.equal(serverName, 'serverName');
-		assert.equal(databaseName, 'databaseName');
+	test('getClusterEndpoints Test', async function (): Promise<void> {
+		let serverInfo = <ServerInfo>{
+			options: {}
+		};
+
+		serverInfo.options[clusterEndpointsProperty] = undefined;
+		let result = getClusterEndpoints(serverInfo);
+		assert.deepStrictEqual(result, []);
+
+		serverInfo.options[clusterEndpointsProperty] = [];
+		result = getClusterEndpoints(serverInfo);
+		assert.deepStrictEqual(result, []);
+
+		let testEndpoint = <RawEndpoint>{
+			serviceName: 'testName',
+			description: 'testDescription',
+			endpoint: 'testEndpoint',
+			protocol: 'testProtocol',
+			ipAddress: 'testIpAddress',
+			port: 1433
+		};
+		serverInfo.options[clusterEndpointsProperty] = [testEndpoint];
+		result = getClusterEndpoints(serverInfo);
+		assert.deepStrictEqual(result, [<IEndpoint>{
+			serviceName: testEndpoint.serviceName,
+			description: testEndpoint.description,
+			endpoint: testEndpoint.endpoint,
+			protocol: testEndpoint.protocol
+		}]);
+
+		testEndpoint.endpoint = undefined;
+		result = getClusterEndpoints(serverInfo);
+		assert.deepStrictEqual(result, [<IEndpoint>{
+			serviceName: testEndpoint.serviceName,
+			description: testEndpoint.description,
+			endpoint: 'https://testIpAddress:1433',
+			protocol: testEndpoint.protocol
+		}]);
 	});
 
-	test('Should extract server and database name with other parentheses', async function (): Promise<void> {
-		let serverName = getServerFromFormattedAttachToName('serv()erName (databaseName)');
-		let databaseName = getDatabaseFromFormattedAttachToName('serv()erName (databaseName)');
-		assert.equal(serverName, 'serv()erName');
-		assert.equal(databaseName, 'databaseName');
+	test('getHostAndPortFromEndpoint Test', async function (): Promise<void> {
+		let result = getHostAndPortFromEndpoint('https://localhost:1433');
```

Comment:
just for sanity can we include some ports that aren't 4 digits long as well?

Reply:
Done